spec.html

```diff
@@ -24176,16 +24314,265 @@ <h1>Runtime Semantics: Evaluation</h1>
+          _extraInitializers_: a List of function objects,
+          _isStatic_: a Boolean,
+        ): either a normal completion containing ~unused~, or an abrupt completion
+      </h1>
+      <dl class="header">
+        <dt>description</dt>
+        <dd>Applies the decorators stored on an ClassElementDefinition to the value of the definition.</dd>
+      </dl>
+      <emu-alg>
+        1. Let _decorators_ be _elementRecord_.[[Decorators]].
+        1. If _decorators_ is ~empty~, return ~unused~.
+        1. Let _key_ be _elementRecord_.[[Key]].
+        1. Let _kind_ be _elementRecord_.[[Kind]].
+        1. For each element _decorator_ of _decorators_, do
+          1. Let _decorationState_ be the Record { [[Finished]]: *false* }.
+          1. Let _context_ be CreateDecoratorContextObject(_kind_, _key_, _extraInitializers_, _decorationState_, _isStatic_).
+          1. Let _value_ be *undefined*.
+          1. If _kind_ is ~method~, set _value_ to _elementRecord_.[[Value]].
+          1. Else if _kind_ is ~getter~, set _value_ to _elementRecord_.[[Get]].
+          1. Else if _kind_ is ~setter~, set _value_ to _elementRecord_.[[Set]].
+          1. Else if _kind_ is ~accessor~, then
+            1. Set _value_ to OrdinaryObjectCreate(%Object.prototype%).
+            1. Perform ! CreateDataPropertyOrThrow(_accessor_, *"get"*, _elementRecord_.[[Get]]).
+            1. Perform ! CreateDataPropertyOrThrow(_accessor_, *"set"*, _elementRecord_.[[Set]]).
+          1. Let _newValue_ be ? Call(_decorator_, *undefined*, &laquo; _value_, _context_ &raquo;).
+          1. Set _decorationState_.[[Finished]] to *true*.
+          1. If _kind_ is ~field~, then
+            1. If IsCallable(_initializer_) is *true*, append _initializer_ to _elementRecord_.[[Initializers]].
+            1. Else if _initializer_ is not *undefined*, throw a *TypeError* exception.
+          1. Else if _kind_ is ~accessor~, then
+            1. If _newValue_ is an Object, then
+              1. Let _newGetter_ be ? Get(_newValue_, *"get"*).
+              1. If IsCallable(_newGetter_) is *true*, set _elementRecord_.[[Get]] to _newGetter_.
+              1. Else if _newGetter_ is not *undefined*, throw a *TypeError* exception.
+              1. Let _newSetter_ be ? Get(_newValue_, *"set"*).
+              1. If IsCallable(_newSetter_) is *true*, set _elementRecord_.[[Set]] to _newSetter_.
+              1. Else if _newSetter_ is not *undefined*, throw a *TypeError* exception.
+              1. Let _initializer_ be ? Get(_newValue_, *"init"*).
+              1. If IsCallable(_initializer_) is *true*, append _initializer_ to _elementRecord_.[[Initializers]].
+              1. Else if _initializer_ is not *undefined*, throw a *TypeError* exception.
+            1. Else if _newValue_ is not *undefined*, throw a *TypeError* exception.
+          1. Else,
+            1. If IsCallable(_newValue_) is *true*, then
+              1. Perform MakeMethod(_newValue_, _homeObject_).
```

Comment:
Note that this call will fail if <var>newValue</var> doesn’t have a [[HomeObject]] internal slot, which would be the case when <var>newValue</var> is a built‑in function, which is now also the case for a class that’s using a default constructor[^1].

[^1]: <https://github.com/tc39/ecma262/pull/2216>

---

In fact, the [[HomeObject]] should probably not be changed at all if it’s not `undefined`, otherwise it will cause weirdness, especially when <var>newValue</var> is a method taken from an existing class.

Reply:
This behavior is being discussed here as well: https://github.com/tc39/proposal-decorators/issues/497

I have it on the agenda for the next plenary to remove this line, as it does result in very odd behavior that does not seem desirable.